### DISCUSSION TITLE
How to config jenkinsfile's kubernetes gradle version ?

### DISCUSSION THREAD

**Discussion by Break4ever** [thyrlian/AndroidSDK#70] [Q&A] [answered]
The following jenkinsfile is my part of example:

containers: [
        containerTemplate(name: 'android-env', image: 'thyrlian/android-sdk:latest', ttyEnabled: true, command: 'cat', alwaysPullImage: false),
        containerTemplate(name: 'docker', image: 'docker:19.03.1-dind', ttyEnabled: true, command: 'cat'),
        containerTemplate(name: 'curl', image: 'byrnedo/alpine-curl:0.1.8', ttyEnabled: true, command: 'cat')
    ]

How to config gradle version and kotlin version for 'thyrlian/android-sdk:latest' image ? @thyrlian

**Comment by thyrlian** [comment, marked as answer]
Hi @Break4ever.

If I'm not mistaken, in the given example you use Kubernetes plugin, sorry that I know nothing about it.  I understand that you would like to have specific versions of Gradle or Kotlin used in your Android build environment, that's not a problem, but not exactly the way you planned.

We don't provide Docker images with all versions of tools, for below reasons:
* According to Android's [terms and conditions](https://developer.android.com/studio/terms), one may not redistribute the SDK or any part of the SDK.  This is mentioned in the [README](https://github.com/thyrlian/AndroidSDK#philosophy).  Thus, we're supposed to provide only a skeleton of an environment.
* There are a bunch of other supporting tools, such as Java, Kotlin, Gradle, a developer/team can use any combination of different versions of those tools based on the requirement, we can't build them all to satisfy everyone.

Having that said, it's just a matter that you can't pull any existing images, but no one can stop you from customizing and building your own image.  It's actually fairly simple.  From the [README](https://github.com/thyrlian/AndroidSDK#getting-started), the building command is:

```bash
docker build --build-arg JDK_VERSION=<jdk_version> --build-arg GRADLE_VERSION=<gradle_version> --build-arg KOTLIN_VERSION=<kotlin_version> --build-arg ANDROID_SDK_VERSION=<android_sdk_version> -t android-sdk android-sdk
```

I've checked Kubernetes plugin's [guide](https://www.jenkins.io/doc/pipeline/steps/kubernetes/), and it seems there is only an option of pulling an image, but not building.  However, with the [Docker Pipeline](https://plugins.jenkins.io/docker-workflow/), you can build on your own:

```bash
docker.build("my-image:${env.BUILD_ID}", "--build-arg GRADLE_VERSION=<gradle_version> --build-arg KOTLIN_VERSION=<kotlin_version> -f Dockerfile .")
```

Hope this helps.

**Reply by Break4ever** [reply]
Thanks for your patient and help !

**Reply by thyrlian** [reply]
Anytime 😄
